### PR TITLE
Vault config store

### PR DESCRIPTION
**Pull request by cescoffier**
This PR provides the support for Vault (https://www.vaultproject.io/). This config store retrieves secrets from Vault.

Are supported:
* direct token
* token generation
* userpass (username password) backend
* approle backend
* certificates (certs backend)

Renewable tokens are renewed automatically. If not renewable, authentication is re-executed upon expiration of the token.

### CODE REVIEW

File: vertx-config-vault/src/main/java/io/vertx/config/vault/VaultConfigStore.java [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+  @Override
+  public void close(Handler<Void> completionHandler) {
+    client.close();
+    completionHandler.handle(null);
+  }
+
+  @Override
+  public void get(Handler<AsyncResult<Buffer>> completionHandler) {
+    if (context == null) {
```
Comment (ruslansennov) [inline vertx-config-vault/src/main/java/io/vertx/config/vault/VaultConfigStore.java]: well, it seems I should add the same into `vertx-config-consul` module and keep `ConsulClient` as is...

File: vertx-config-vault/src/main/java/examples/Examples.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+      .setType("vault")
+      .setConfig(config);
+
+    ConfigRetriever retriever = ConfigRetriever.create(vertx,
+      new ConfigRetrieverOptions().addStore(store));
+  }
+
+  public void example1WithConfig(Vertx vertx) {
+    JsonObject vault_config = new JsonObject()
+      .put("host", "1270.0.1") // The host name
```
Comment (ruslansennov) [inline vertx-config-vault/src/main/java/examples/Examples.java]: 127.0.0.1

File: vertx-config-vault/src/main/java/io/vertx/config/vault/client/SlimVaultClient.java [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+  }
+
+  /**
+   * Lists secrets from path (children).
+   *
+   * @param path          the path
+   * @param resultHandler the callback invoked with the result
+   */
+  public void list(String path, Handler<AsyncResult<List<String>>> resultHandler) {
+    String fullPath = path == null ? "list=true" : path + "?list=true";
```
Comment (ruslansennov) [inline vertx-config-vault/src/main/java/io/vertx/config/vault/client/SlimVaultClient.java]: if `list=true` is a GET parameters, it should be preceded with `?` even if `path == null`

Comment (cescoffier) [inline vertx-config-vault/src/main/java/io/vertx/config/vault/client/SlimVaultClient.java]: You are absolutely right. Actually according to https://www.vaultproject.io/api/secret/generic/index.html#list-secrets the path is required.

File: vertx-config-vault/src/main/java/io/vertx/config/vault/VaultConfigStore.java (hunk 1 above)
Comment (cescoffier) [inline vertx-config-vault/src/main/java/io/vertx/config/vault/VaultConfigStore.java]: Yes.

**Comment by cescoffier** [comment]
Thanks @ruslansennov

**Pull request merged**